core/src/test/java/org/kiwix/kiwixmobile/core/utils/ExternalLinkOpenerTest.kt

```diff
@@ -34,48 +34,77 @@ import org.kiwix.kiwixmobile.core.extensions.toast
 
 internal class ExternalLinkOpenerTest {
   private val sharedPreferenceUtil: SharedPreferenceUtil = mockk()
-  private val alertDialogShower: AlertDialogShower = mockk()
+  private val alertDialogShower: AlertDialogShower = mockk(relaxed = true)
   private val intent: Intent = mockk()
   private val activity: Activity = mockk()
 
   @Test
-  internal fun `requestOpenLink is called if external link popup preference is true`() {
+  internal fun `alertDialogShower opens link if confirm-button is clicked`() {
     every { intent.resolveActivity(activity.packageManager) } returns mockk()
     every { sharedPreferenceUtil.prefExternalLinkPopup } returns true
+    val lambdaSlot = slot<() -> Unit>()
     val externalLinkOpener =
-      spyk(
-        ExternalLinkOpener(activity, sharedPreferenceUtil, alertDialogShower),
-        recordPrivateCalls = true
-      )
+      spyk(ExternalLinkOpener(activity, sharedPreferenceUtil, alertDialogShower))
     externalLinkOpener.openExternalUrl(intent)
-    verify { externalLinkOpener["requestOpenLink"](intent) }
+    verify {
+      alertDialogShower.show(any(), capture(lambdaSlot), any(), any())
+    }
+    lambdaSlot.captured.invoke()
+    verify { activity.startActivity(intent) }
   }
 
   @Test
-  internal fun `openLink is called if external link popup preference is false`() {
+  internal fun `alertDialogShower does not open link if negative-button is clicked`() {
+    every { intent.resolveActivity(activity.packageManager) } returns mockk()
+    every { sharedPreferenceUtil.prefExternalLinkPopup } returns true
+    val lambdaSlot = slot<() -> Unit>()
+    val externalLinkOpener =
+      spyk(ExternalLinkOpener(activity, sharedPreferenceUtil, alertDialogShower))
+    externalLinkOpener.openExternalUrl(intent)
+    verify {
+      alertDialogShower.show(any(), any(), capture(lambdaSlot), any())
+    }
+    lambdaSlot.captured.invoke()
+    verify(exactly = 0) { activity.startActivity(intent) }
```

Comment:
I am not sure what my preferred syntax for this situation is

you can `verify { ... } wasNot Called`

you can use `confirmVerified` which checks every call on a mock was declared in a `verify` and equally in that scenario you can use `verifyAll/Sequence` which has to have all the calls to all mocks.

probably `wasNot Called` is enough here

Reply:
`verify { activity.startActivity(intent) wasNot Called }` does not work for some reason? I get `java.lang.AssertionError: Verification failed: call 1 of 1: Activity(#4).startActivity(eq(Intent(#3)))) was not called.`

Reply:
Is an empty `confirmVerified()` the same as guaranteeing nothing was called?

Reply:
I am actually not sure where `wasNot Called` is supposed to go, I'll go to an IDE and figure some stuff out.

`confirmVerified` operates on a mock, empty is meaningless with it. for the mock you pass it it checks every interaction with the mock has a `verify` call. this means if this test execution actually had invoked `activity.start` then the test would fail because it was not declared verified

Reply:
![image](https://user-images.githubusercontent.com/3358282/88903417-9278ec80-d24b-11ea-9a47-1f30a081dccf.png)

Reply:
lets just keep `exactly=0` this rabbit hole got a bit too deep

Reply:
haha, ok nice